### ISSUE TITLE
Acme signer reissue a certificate if private key isn't RSA

### ISSUE THREAD

**Issue by pbabilas** [jcmoraisjr/haproxy-ingress#684]
**Description of the problem**

My ingress configuration is configured to work with acme cert generation, but when I set my own client custom cert I would like to not overwrite the client cert with new LE cert.
I've done tests and after apply new tls secret, haproxy seams to rewrite client cert with LE one. Can't find any reason of that :(

```I1021 07:35:32.150444       6 instance.go:102] starting certificate check (started leading)
I1021 07:35:32.150719       6 instance.go:110] finish adding 65 certificate(s) to the work queue
I1021 07:35:32.164514       6 signer.go:142] acme: authorizing: id=1 secret=aps/aps-tls domain(s)=apspolska.pl,www.apspolska.pl endpoint=https://acme-v02.api.letsencrypt.org reason='certificate does not exist'
I1021 07:35:35.134830       6 signer.go:147] acme: new certificate issued: id=1 secret=aps/aps-tls domain(s)=apspolska.pl,www.apspolska.pl
I1021 07:35:35.137018       6 backend_ssl.go:57] updating secret aps/aps-tls in the local store
```

**Expected behavior**

When I set custom cert, ingres should not rewrite this cert with LE cert.

**Steps to reproduce the problem**

1. apply any other cert to secret
2. wait until refresh is started
3. check secret was changed

**Environment information**

HAProxy Ingress version: `v0.11-snapshot.3`

Command-line options:

```
--configmap=ingress-controller/haproxy-ingress
--acme-server
--acme-track-tls-annotation
```

Global options:

```
acme-emails: pawel.babilas@merce.com
  acme-endpoint: v2
  acme-terms-agreed: "true"
  dynamic-scaling: "false"
  max-connections: "50000"
  prometheus-port: "9105"
  ssl-redirect-code: "301"
  use-proxy-protocol: "true"

```

**Comment by pbabilas** [comment]
ping, anyone?

**Comment by jcmoraisjr** [comment]
Hi, I'm working in some doc related stuff these days and could have a closer look into this issue. I'll ping back shortly.

**Comment by pbabilas** [comment]
Thansk, for reply.
More details:

`I1022 07:32:53.641110       6 signer.go:142] acme: authorizing: id=5 secret=aps/aps-tls domain(s)=apspolska.pl,www.apspolska.pl endpoint=https://acme-v02.api.letsencrypt.org reason='certificate does not exist'
I1022 07:32:56.619048       6 signer.go:147] acme: new certificate issued: id=5 secret=aps/aps-tls domain(s)=apspolska.pl,www.apspolska.pl
`

It appears that ingress can not find tls in secret. But tls exists and working via http. and curl ;]

**Comment by jcmoraisjr** [comment]
Hi, according to the provided log, the certificate was issued due to one of the following reasons:

* the provided secret does not exist;
* the `tls.crt` or the `tls.key` keys does not exist in the provided secret;
* the crt or the key couldn't be base64 decoded;
* the decoded der certificate or key couldn't be parsed as a valid x509 crt/key pair

This is the relevant piece of code that does such validations. Perhaps your certificate has eg a pretty new extension that the go version being used doesn't recognize?

https://github.com/jcmoraisjr/haproxy-ingress/blob/67f1bdcf2efb4be44bb3c13cba89ee2a473ab1f9/pkg/controller/cache.go#L447-L471

**Comment by pbabilas** [comment]
Hi, thanks for answer.
I've already check the source code, ald everything seems to be ok. Secret exists, tls.crt and tls.key also are present.
both can be decoded by base64, and keys are velid crt/key pair:

```
 b-code@xps  ~/workspace/certs  openssl x509 -noout -modulus -in apspolska.pl.pem|openssl md5                                                                                                                         ✔  11963  16:54:34
(stdin)= c8aa29ac3210ee12defb913123e58f3a
 b-code@xps  ~/workspace/certs  openssl rsa -noout -modulus -in apspolska.pl.key |openssl md5                                                                                                                         ✔  11964  16:55:43
(stdin)= c8aa29ac3210ee12defb913123e58f3a
```
All online cert decoders can decode cert correctly.
Dunno how to debug this issue. And its not only one cert. Only one is not rewritable. Certs is from different cert distributor, ald all are always rewrite. Can you give me any hint ho to debug this issue by my own?

**Comment by pbabilas** [comment]
certificates are in v3:
```
Certificate:
    Data:
        Version: 3 (0x2)
```
Both the one that is replaced and the one that is working...

**Comment by jcmoraisjr** [comment]
In order to debug this you can either:

* Change and deploy your own version of the controller. `glog` the `GetTLSSecretContent` in order to know which call is broken and why;
* Write a small go code that tries to decode and parse the certificate.

I'll also add some verbosity in this code so one can see a proper reason in the future.

**Comment by pbabilas** [comment]
OK, I've just used Your piece of code for verbosity and now gets that error:

```
reason='certificate does not exist (error parsing key of secret aps/aps-tls: x509: failed to parse private key (use ParsePKCS8PrivateKey instead for this key format))'
```

So it means that go lib can not parse my certs?

**Comment by jcmoraisjr** [comment]
No, this means that you don't have a RSA private key. It seems that this piece of code doesn't need a parsed private key at all and we can fix this by removing the parsing. Thanks for the test.

**Comment by pbabilas** [comment]
ok, gr8 I will check if convertin pk to RSA will resolve my issue.

**Comment by pbabilas** [comment]
after change key to rsa my cert is still rewrited but only logs i've get is:

```
I1103 07:32:42.066532       6 signer.go:142] acme: authorizing: id=27 secret=aps/aps-tls domain(s)=apspolska.pl,www.apspolska.pl endpoint=https://acme-v02.api.letsencrypt.org reason='added one or more domains to an existing certificate'
I1103 07:32:48.018550       6 signer.go:147] acme: new certificate issued: id=27 secret=aps/aps-tls domain(s)=apspolska.pl,www.apspolska.pl
```

**Comment by pbabilas** [comment]
Maybe match method is not working correctly.
My cert has:

Subject Alternative Names: *.apspolska.pl, apspolska.pl

So it seams to be ok ;>

**Comment by jcmoraisjr** [comment]
Yep, the current implementation does a simple match without checking if a wildcard domain includes another one. I'll take care of this as well, but please file this on another issue so this won't be forgotten. Thank you for the tests so far.

**Comment by pbabilas** [comment]
ok, I've made PR for this issue:
https://github.com/jcmoraisjr/haproxy-ingress/pull/693

**Comment by jcmoraisjr** [comment]
v0.11-beta.4 should have solved this one, primary key isn't parsed anymore. Merged to v0.10 as well, release in a few days. Closing.
